include/envoy/thread/thread.h

```diff
@@ -2,40 +2,149 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_ != nullptr) {
+      lock_->lock();
+    }
+  }
+
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
+    if (lock_ != nullptr) {
+      lock_->unlock();
+    }
+  }
+
+private:
+  BasicLockable* const lock_;
+};
+
+/**
+ * Like LockGuard, but uses a tryLock() on construction rather than a lock().
+ */
+class SCOPED_LOCKABLE TryLockGuard {
 public:
-  OptionalLockGuard(T* lock) : lock_(lock) {
-    if (lock) {
-      lock->lock();
+  /**
+   * Establishes a scoped mutex-lock; the a mutex lock is attempted via tryLock, so
+   * an expected outcome is that the lock may fail. isLocked() must be called to
+   * determine whether he lock was actually acquired.
+   *
+   * @param lock the mutex.
+   */
+  TryLockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock)
```

Comment:
I agree; that's the intuitive pattern, and I tried it earlier.  This is what you get:

```
ERROR: /usr/local/google/home/jmarantz/git3/envoy/source/server/BUILD:107:1: C++ compilation of rule '//source/server:hot_restart_lib' failed (Exit 1)
source/server/hot_restart_impl.cc:267:1: error: mutex 'lock' is not held on every path through here [-Werror,-Wthread-safety-analysis]
}
^
source/server/hot_restart_impl.cc:260:12: note: mutex acquired here
  if (lock.tryLock()) {
           ^
source/server/hot_restart_impl.cc:267:1: error: releasing mutex 'lock' that was not held [-Werror,-Wthread-safety-analysis]
}
```
line 267 is the closing brace of HotRestartImpl::getParentStats()

Reply:
I think I'd rather omit the annotations for this class than use incorrect/misleading annotations. How much do we lose if we don't have an annotated tryLock wrapper?

Reply:
We'd lose nothing in the codebase, since there are no GUARDED_BY variables in hot_restart_impl.  I'd have to remove or comment out the new TryLockGuard test I added.

Reply:
I did some investigation, and I think I found a path that can work, and is annotated correctly. Basically, I don't think there's any way to have an EXCLUSIVE_TRYLOCK_FUNCTION on a SCOPED_LOCKABLE, because there's no way to annotate the conditional unlocking for the destructor (among other problems). 

But I think it will work to throw from the constructor if tryLock() returns false, and the annotations are happy with that. Something like the following (but needs cleaning up).

```
diff --git a/include/envoy/thread/thread.h b/include/envoy/thread/thread.h
index 4148caf..227513b 100644
--- a/include/envoy/thread/thread.h
+++ b/include/envoy/thread/thread.h
@@ -61,25 +61,19 @@ public:
    *
    * @param lock the mutex.
    */
-  TryLockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock)
-      : lock_(lock.tryLock() ? &lock : nullptr) {}
-
-  /**
-   * Destruction of the DeferredLockGuard unlocks the lock, if it was locked.
-   */
-  ~TryLockGuard() UNLOCK_FUNCTION() {
-    if (isLocked()) {
-      lock_->unlock();
+  TryLockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (!lock_.tryLock()) {
+      throw std::exception();
     }
   }
 
   /**
-   * @return bool whether the lock was successfully acquired.
+   * Destruction of the DeferredLockGuard unlocks the lock, if it was locked.
    */
-  bool isLocked() const { return lock_ != nullptr; }
+  ~TryLockGuard() UNLOCK_FUNCTION() { lock_.unlock(); }
 
 private:
-  BasicLockable* const lock_;
+  BasicLockable& lock_;
 };
 
 /**
diff --git a/source/server/hot_restart_impl.cc b/source/server/hot_restart_impl.cc
index 1560a6d..3b705a1 100644
--- a/source/server/hot_restart_impl.cc
+++ b/source/server/hot_restart_impl.cc
@@ -256,13 +256,15 @@ void HotRestartImpl::getParentStats(GetParentStatsInfo& info) {
     return;
   }
 
-  Thread::TryLockGuard lock(init_lock_);
-  if (lock.isLocked()) {
+  try {
+    Thread::TryLockGuard lock(init_lock_);
     RpcBase rpc(RpcMessageType::GetStatsRequest);
     sendMessage(parent_address_, rpc);
     RpcGetStatsReply* reply = receiveTypedRpc<RpcGetStatsReply, RpcMessageType::GetStatsReply>();
     info.memory_allocated_ = reply->memory_allocated_;
     info.num_connections_ = reply->num_connections_;
+  } catch (const EnvoyException&) {
+    // Failed to tryLock(); ignore.
   }
 }
 
diff --git a/test/common/common/thread_test.cc b/test/common/common/thread_test.cc
index b211115..3c9b393 100644
--- a/test/common/common/thread_test.cc
+++ b/test/common/common/thread_test.cc
@@ -31,9 +31,7 @@ TEST_F(ThreadTest, TestReleasableLockGuard) {
 
 TEST_F(ThreadTest, TestTryLockGuard) {
   TryLockGuard lock(a_mutex_);
-  if (lock.isLocked()) {
-    EXPECT_EQ(1, ++a_);
-  }
+  EXPECT_EQ(1, ++a_);
 }
 
 } // namespace Thread
```